### PR TITLE
modify reconciler to use new Delta() compare

### PR DESCRIPTION
**Pull request by jaypipes**
Updates the ACK runtime and type system to use the new
`pkg/compare.Delta` comparison functionality. The
`AWSResourceDescriptor.Diff()` and `AWSResourceDescriptor.Equal()`
methods were removed and replaced with a single
`AWSResourceDescriptor.Delta()` method that returns a
`*pkg/compare.Delta` struct containing the difference between two
objects under comparison.

A [corresponding patch](https://github.com/aws-controllers-k8s/code-generator/pull/29) to the
github.com/aws-controllers-k8s/code-generator repository adds the
generation of the `newResourceDelta()` resource package-local function
that recurses into an object and compares the fields in the objects to
each other.

Included in this patch is an update to the `mocks` Makefile target which
now builds mocks for `k8s.io/apimachinery` interfaces that we use in a
new unit test of the update code path in the reconciler.

In order to adhere to best practices of only using exported functions in
Go module testing, I ended up exporting the
`pkg/runtime.reconciler.sync()` method as a `Sync()` method on the
`AWSResourceReconciler` interface. The unit test features a whole lot of
mock setups and definitely could use some refactoring to reduce the
number of setup mocks needed.

IMPORTANT! This is a BREAKING change to the runtime interfaces. Once
merged, we will need to cut a new Git release to signal to consumers
that this is a backwards-incompatible change and if they attempt to
regenerate their controllers with the new code and their controllers
rely on the older interfaces that uses a "diffReporter" struct, they
will generate broken code.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by vijtrip2 — APPROVED** [review]
LGTM!

**Review by RedbackThomson — APPROVED** [review]
I look forward to rebasing on top of this. LGTM

File: pkg/runtime/reconciler_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	// `AWSResourceDescriptor.Delta()` returned a non-empty Delta, that we end
+	// up calling the AWSResourceManager.Update() call in the Reconciler.Sync()
+	// method,
+	r := ackrt.NewReconciler(sc, rmf, fakeLogger, cfg, metrics)
+
+	err := r.Sync(ctx, rm, desired)
+	require.Nil(err)
+	rm.AssertCalled(t, "ReadOne", ctx, desired)
+	rd.AssertCalled(t, "Delta", desired, latest)
+	rm.AssertCalled(t, "Update", ctx, desired, latest, delta)
```
Comment (A-Hilaly) [inline pkg/runtime/reconciler_test.go:131]: I really like this! In the future we can write more unit test for different reconciliation cases

File: mocks/ctrlrt/client.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
+// Code generated by mockery v2.2.2. DO NOT EDIT.
```
Comment (A-Hilaly) [inline mocks/ctrlrt/client.go:1]: +1 for generating mocks for the `ctrlrt/client`, it looks like the fake client is far from being complete https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/fake/doc.go#L27-L36

**Review by surajkota — APPROVED** [review]

**Comment by surajkota** [comment]
It will be good to have some details regarding how this PR has been tested

**Comment by jaypipes** [comment]
> It will be good to have some details regarding how this PR has been tested

@surajkota yeah, agreed. basically, what I did is rebuild the S3 controller after creating a local go.mod file and ran `make local-test`

File: Makefile [hunk 3]
```diff
@@ -20,6 +22,15 @@ mocks: install-mockery ## Build mocks
...
 	@echo "ok."
+	@echo -n "building mocks for k8s.io/apimachinery/pkg/apis/meta/v1 ... "
+	@bin/mockery --quiet --name=Object --case=underscore --output=mocks/apimachinery/pkg/apis/meta/v1 --dir="$(K8S_APIMACHINERY_DIR)/pkg/apis/meta/v1"
+	@echo "ok."
+	@echo -n "building mocks for k8s.io/apimachinery/runtime ... "
+	@bin/mockery --quiet --name=Object --case=underscore --output=mocks/apimachinery/pkg/runtime --dir="$(K8S_APIMACHINERY_DIR)/pkg/runtime"
+	@echo "ok."
+	@echo -n "building mocks for k8s.io/apimachinery/runtime/schema ... "
+	@bin/mockery --quiet --name=ObjectKind --case=underscore --output=mocks/apimachinery/pkg/runtime/schema --dir="$(K8S_APIMACHINERY_DIR)/pkg/runtime/schema"
+	@echo "ok."
```
[the remark this answers is not in the document]
Comment (jaypipes) [reply Makefile:33]: > Not sure but maybe the command used to generate`mocks/ctrlrt/*` is missing here?

Nah, I accidentally left the mocks/ctrlrt files in there. Fixed.

File: Makefile [hunk 4]
```diff
@@ -1,7 +1,9 @@
 SHELL := /bin/bash # Use bash syntax
 
-# Set up variables
+GOPATH ?= "$(HOME)/go"
 GO111MODULE=on
+K8S_APIMACHINERY_VERSION = $(shell grep k8s.io/apimachinery go.mod | cut -d' ' -f 2 | tr -d '[:space:]')
```
[the remark this answers is not in the document]
Comment (jaypipes) [reply Makefile:5]: ooh, cool. :) changed.

**Pull request merged**